src/BinSkim.Sdk/Telemetry.cs

```diff
@@ -0,0 +1,183 @@
+    /// If Application Insights is enabled, then the <see cref="TelemetryClient"/>
+    /// property will be non-null.
+    /// </summary>
+    public sealed class Telemetry : IDisposable
+    {
+        /// <summary>
+        /// The name of an environment variable that holds the connection
+        /// string for an Application Insights resource.
+        /// </summary>
+        internal const string AppInsightsConnectionStringEnvVar = "BinskimAppInsightsConnectionString";
+
+        /// <summary>
+        /// The name of an environment variable that holds the instrumentation
+        /// key for an Application Insights resource.
+        /// </summary>
+        internal const string AppInsightsInstrumentationKeyEnvVar = "BinskimCompilerDataAppInsightsKey";
+
+        /// <summary>
+        /// Gets the telemetry configuration. May be null if Application Insights
+        /// telemetry is not enabled.
+        /// </summary>
+        private TelemetryConfiguration? TelemetryConfiguration { get; }
+
+        /// <summary>
+        /// Gets the telemetry client. This may be null if Application Insights
+        /// telemetry is not enabled.
+        /// </summary>
+        public TelemetryClient? TelemetryClient { get; }
+
+        /// <summary>
+        /// Construct a new <see cref="Telemetry"/> instance from environment variables.
+        /// </summary>
+        /// <remarks>
+        /// If environment variables aren't set, then <see cref="TelemetryClient"/> will be null.
+        /// </remarks>
+        public Telemetry() : this(CreateTelemetryConfigurationFromEnvironment())
+        {
+        }
+
+        /// <summary>
+        /// Construct a new <see cref="Telemetry"/> instance using
+        /// the given <see cref="ApplicationInsights.Extensibility.TelemetryConfiguration"/>.
+        /// </summary>
+        /// <param name="telemetryConfiguration">The pipeline to use. May be null, in which case
+        /// Application Insights is disabled.</param>
+        /// <remarks>
+        /// If <paramref name="telemetryConfiguration"/> is non-null, it will be
+        /// disposed when this object is disposed.
+        /// </remarks>
+        internal Telemetry(TelemetryConfiguration? telemetryConfiguration)
+        {
+            if (telemetryConfiguration != null)
+            {
+                var telemetryClient = new TelemetryClient(telemetryConfiguration);
+                ConfigureTelemetryContext(telemetryClient.Context);
+                this.TelemetryClient = telemetryClient;
+                this.TelemetryConfiguration = telemetryConfiguration;
+            }
+        }
+
+        /// <inheritdoc/>
+        public void Dispose()
+        {
+            TelemetryClient? telemetryClient = TelemetryClient;
+            if (telemetryClient is null)
+            {
+                return;
+            }
+
+            // Flush and close AppInsights client.
+            telemetryClient.Flush();
```

Comment:
>telemetryClient.Flush(); [](http://example.com/codeflow?start=12&length=24)

Looks like there is an async version `[FlushAsync]([https://github.com/microsoft/ApplicationInsights-dotnet/pull/2202)`](https://github.com/microsoft/ApplicationInsights-dotnet/pull/2202)`) merged in April 2021. Not sure if its released

Reply:
There is, but we wouldn't want it here. We want a **synchronous** flush (a method that blocks until all buffered events are transmitted). Otherwise, this would have to be an async method and IAsyncDisposable.

Reply:
ok do you mean even [FlushAsync] may return before all telemetries are sent?

I saw below comment at this closed [issue](https://github.com/MicrosoftDocs/azure-docs/issues/98256)

>>>Any user who wants to avoid the Sleep should shut down their application using FlushAsync instead.

But the Azure Doc page still says sleep is required.

Reply:
Because it simplifies the code (and "hides" the problem), I'll switch to FlushAsync.